lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+                distinct from other instances that may be hanlding other events.
+            func_run: a closure to run when we have a lock. (It must take a CharmBase object and
+                EventBase object as args.)
+            event_acquire: The event that will prompt us to acquire a lock.
+        """
+        super().__init__(charm, None)
+
+        self.charm = charm
+        self.name = rel_name
+        self._func_run = func_run
+
+        self.framework.observe(event_acquire, self._on_acquire_lock)
+        self.framework.observe(charm.on[self.name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on.run_with_lock, self._on_run_with_lock)
+        self.framework.observe(charm.on.process_locks, self._on_process_locks)
+
+    def _func_run(self: CharmBase, event: EventBase) -> None:
+        """Placeholder for the function that actually runs our event.
+
+        Usually overridden in the init.
+        """
+        raise NotImplementedError
+
+    def _on_relation_changed(self: CharmBase, event: RelationChangedEvent):
+        """Process relation changed.
+
+        First, determine whether this unit has been granted a lock. If so, emit a RunWithLock
+        event.
+
+        Then, if we are the leader, fire off a process locks event.
+
+        """
+        if not event.relation.name == self.name:
+            # This is not relevant to us.
+            return
+
+        if Lock(self.name, self.charm).is_held():
+            self.charm.on.run_with_lock.emit(name=self.name)
+
+        if self.framework.model.unit.is_leader():
+            self.charm.on.process_locks.emit(name=self.name)
+
+    def _on_process_locks(self: CharmBase, event: ProcessLocks):
+        """Process locks.
+
+        Runs only on the leader. Updates the status of all locks.
+
+        """
+        if not event.name == self.name:
+            return
+
+        if not self.framework.model.unit.is_leader():
+            return
+
+        pending = []
+
+        for lock in Locks(self.name, self.charm):
+            if lock.is_held():
+                # One of our units has the lock -- return without further processing.
+                return
+
+            if lock.release_requested():
+                lock.clear()  # Updates relation data
+
+            if lock.is_pending():
+                if lock.unit == self.charm.model.unit:
+                    # Always run on the leader last.
+                    pending.insert(0, lock)
+                else:
+                    pending.append(lock)
+
+        # If we reach this point, and we have pending units, we want to grant a lock to
+        # one of them.
+        if pending:
+            lock = pending[-1]
+            lock.grant()
+            if lock.unit == self.charm.model.unit:
+                # It's time for the leader to run with lock.
+                self.charm.on.run_with_lock.emit(name=self.name)
```

Comment:
Is a relation_changed event not fired for a unit when it modifies its own relation data?

Reply:
In the test harness, it does not.

In production, it might -- that's one of the things that I need to test before this becomes production worthy.